### PR TITLE
Add transient settings to Topic QoS (#295).

### PR DESCRIPTION
**Pull request by martinbremmer**
When data should be transient, the Topic QoS should reflect that.
Not only does the Durability need to be set, but also the DurabilityService policy.

See for more info #295.

### CODE REVIEW

**Comment by wjwwood** [comment]
CI (opensplice only):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8873/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4663/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7267/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8781/)

File: rmw_opensplice_cpp/src/qos.cpp
Comment (wjwwood) [inline rmw_opensplice_cpp/src/qos.cpp]: nitpick: this looks like an arbitrary amount of indention

I'm not sure if this package is running our style linter or not (CI will complain if it is), but I'd expect something more like:

```c++
  if (!rmw_qos_to_history(
    qos_profile,
    entity_qos.history.kind,
    entity_qos.history.depth))
  {
    return false;
  }
```

Again, a nitpick and only needs to be addressed if the linter is turned on and it complains.

File: rmw_opensplice_cpp/src/qos.cpp
Comment (wjwwood) [inline rmw_opensplice_cpp/src/qos.cpp]: nitpick: same as above, according to our style guide the indention should only be single steps of (in this case) two spaces

File: rmw_opensplice_cpp/src/qos.cpp
Comment (wjwwood) [inline rmw_opensplice_cpp/src/qos.cpp]: Wouldn't it better to store the result of `check_get_default_topic_qos()` rather than call it twice? Imagine if it was not idempotent (it is now, but if that were changed in the future this code would be broken).

Comment (wjwwood) [reply rmw_opensplice_cpp/src/qos.cpp]: I know you moved this from the other file, but still the comment applies, might as well fix it if we can.

**Review by wjwwood — APPROVED** [review]
change makes sense to me, with some comments/nitpicks

**Comment by martinbremmer** [comment]
Thanks for the response @wjwwood. I just don't know how to interpret the Jenkins results.

**Review by wjwwood — APPROVED** [review]

**Comment by wjwwood** [comment]
Some of the test failures on CI were related to linting, e.g.:

https://ci.ros2.org/job/ci_linux-aarch64/4663/testReport/(root)/projectroot/uncrustify/

Please run the tests locally for the `rmw_opensplice` packages.

**Comment by wjwwood** [comment]
Unfortunately we have some regressions in our CI at the moment so untangling what's broken on master and what's broken here is a bit messy at the moment. I'll have to run CI again and interpret the CI results manually. I'll do that after you run the tests locally and make sure the linters are at least happy, so do let me know when that's done. Hopefully by then CI will be clearer on master.

**Comment by wjwwood** [comment]
`colcon test --packages-select rmw_opensplice_cpp` should let you run at least the tests for that  package.

**Comment by martinbremmer** [comment]
Thanks @wjwwood.
Your ```colcon test --packages-select rmw_opensplice_cpp``` remark cleared up some confusion on my part. I was testing locally, but only ```test_communication```. 

The ```rmw_opensplice_cpp``` tests now pass (locally).

**Comment by wjwwood** [comment]
Ok, here's CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8896/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4683/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7283/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8814/)

**Comment by wjwwood** [comment]
Sorry we're still having issues with CI. I'll have to wait until next week sometime to see if I can merge this.

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8951/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4722/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7324/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8865/)

**Comment by wjwwood** [comment]
OpenSplice only on master for comparison:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8961/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4731/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7333/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8885/)

**Comment by wjwwood** [comment]
It looks like there is still a difference in this branch and master in terms of test failures. Just looking at Linux, for example, this test failed on this branch but not master:

https://ci.ros2.org/job/ci_linux/8951/testReport/junit/(root)/projectroot/test_action_client/

Which actually fails with a segmentation fault. So it seems like this change has an issue somewhere.

**Comment by martinbremmer** [comment]
@wjwwood, the only segmentation faults I could find on https://ci.ros2.org/job/ci_linux/8951/consoleFull are related to ```test_action_client.py``` and ```test_action_server.py```. As far as I can tell, those tests aren't really related to rmw_opensplice. Or am I missing something?

I don't see a problem when running locally:
```
$ python3 ./src/ros2/rclpy/rclpy/test/test_action_client.py 
................
----------------------------------------------------------------------
Ran 16 tests in 9.320s

OK
```

**Comment by dirk-thomas** [comment]
OpenSplice support has been dropped for Foxy so there is no need to iterate on this.

**Comment by clalancette** [comment]
Closing since we no longer support OpenSplice in any active ROS 2 distribution.

**Pull request closed without merging**
